build-system/babel-plugins/babel-plugin-transform-log-methods/index.js

```diff
@@ -0,0 +1,320 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Indirects log messages through expansion calls with
+ * base36-encoded message ids. These redirect to a URL or look up a message
+ * table to interpolate and display the logged string.
+ *
+ *    dev().assert(foo != bar, 'foo should not be bar')
+ *    // transforms into 👇
+ *    dev().assert(foo != bar, dev().expandLogMessage('1z'))
+ *
+ * Where `expandLogMessage` is in charge of looking up and expanding the message
+ * string, or returning a URL where the interpolated message is displayed.
+ *
+ * Additionally outputs a JSON file containing [template, id] pairs keyed by
+ * message template. This table is "backwards" for deduping, to key by id its
+ * rows should be reversed separately.
+ *
+ * When the template has arguments they are nested into the expansion method:
+ *
+ *    dev().assert(false, '%s should not be %s.', foo, bar)
+ *    // transforms into 👇
+ *    dev().assert(false, dev().expandLogMessage('e2', foo, bar))
+ *
+ * It also nests arguments of template literals:
+ *
+ *    dev().assert(false, `Hello, my dear ${name}`);
+ *    // transforms into 👇
+ *    dev().assert(false, dev().expandLogMessage('0a', name));
+ *
+ * The motivation of this transform is to reduce binary size. The average length
+ * of a log message is ~43 whereas the common minified transformed output (as
+ * `x().l('a0')`) is just 11. Since the prefix of the output could repeat
+ * throughout a binary, it also compresses well.
+ *
+ * Resulting compressed, minified binaries reduce size by ~1.5% depending on
+ * their logging density.
+ */
+const fs = require('fs');
+const path = require('path');
+const {
+  messagesPath,
+  singletonFunctions,
+  transformableMethods,
+} = require('../../log-module-metadata.js');
+
+/**
+ * Path to messages file for this version relative to the repo root.
+ * This is a default since tests can configure this table's output path, but for
+ * all building purposes this should be considered the canonical messages
+ * filepath for the built version.
+ */
+const defaultMessagesPath = messagesPath;
+
+/** Method on `Log` that this transform outputs to "expand" log messages. */
+const messageExpansionMethod = 'expandLogMessage';
+
+/**
+ * Mangled expansion method name to prevent large public method name in
+ * multipass build.
+ */
+const messageExpansionMethodShortName = 'l';
+
+/**
+ * Approximate length of a compressed message expansion call to determine
+ * whether an instance of message indirection actually reduces binary size.
+ */
+const roughMinifiedExpansionLength = 'a().b("xx")'.length;
```

Comment:
We could probably make everything shorter if the `assert` method (and the others) did the expansion, instead of having to call `dev().expand()` (`this.` is guaranteed to have an efficient representation in gzip/brotli).